### PR TITLE
Add ADRs to introduce a platform oriented architecture

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

Two ADRs to retrospectively introduce a platform oriented architecture, and to introduce a "UI" architecture module to the core platform.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Review by eskatos — APPROVED** [review]

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
Thanks @eskatos

File: architecture-standards/0004-use-a-platform-architecture.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+This is a platform that builds on the core and software platforms to add support for developing native software.
+
+This platform provides specific support for Swift, C++ and C.
+
+### Cross-cutting architecture modules
+
+There are some additional cross-cutting architecture modules that aren't themselves platforms:
+
+### Enterprise integration
```
Comment (tresat) [inline architecture-standards/0004-use-a-platform-architecture.md:81]: 💅 Shouldn't there be 4 ####'s for the architecture modules?

File: architecture-standards/0005-introduce-core-ui-architecture-module.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+## Context
+
+The Gradle core platform provides many services to the Gradle platforms and build logic. One such group of services allow logic to interact with the build user, to provide diagnostics, progress information, prompt for questions, and so on. Currently, these services are part of the core platform runtime architecture module.
+
+A downside of this structure is that it is difficult to do focussed work on the Gradle UI.
+
+## Decision
+
+Introduce a "UI" architecture module to the core platform, and move the user interaction services to this new module.
```
Comment (tresat) [inline architecture-standards/0005-introduce-core-ui-architecture-module.md:15]: ❓ Where will this module live in the dependency graph of Core architecture modules?  What is the final proposed dependency graph of those modules?

Comment (big-guy) [reply architecture-standards/0005-introduce-core-ui-architecture-module.md:15]:
```
core/runtime <-- core/ui 
                 core/execution <-- (configures) core/configuration
```

**Pull request merged**